### PR TITLE
Fix 16.04 OVA:  isc-dhcp-server can't start

### PR DESCRIPTION
**Pull request by panpan0000**
This is re-work of PR https://github.com/RackHD/RackHD/pull/505

------

**Background:**
   in OVA, the NIC1 is ens33, and in Vagrant , the NIC1 is enp0s8.
   isc-dhcp-server fails to start in 16.04 OVA (because no ens33  in /etc/default/isc-dhcp-server )
   moreover, vagrant 16.04 can't work ,because scripts/addnetif.sh & script/firstuser.sh doesn't consider enp0s8

------

**Root Cause:**
   same as above

------
***Implementation***
    Smartly retrieving the Control Port (Secondary NIC) Name:  leverage "ip addr" command, assume the secondary NIC are listed as index=3 in "ip addr" output.

**Some other fixes **
    1. when /etc/hostname updated, the /etc/hosts remains unchanged (its hostname came from DHCP server). it bring confusion of warning : "unable to resolve host rackhd",
    2. change ```rackhd-demo``` hostname to ```rackhd``` (per @cgx027's comment in PR https://github.com/RackHD/RackHD/pull/505)
    3.  remove ansible/roles/isc-dhcp-server/files/systemd-control.cfg. because it's duplicated with packer/scripts/addnetif.sh, and moreover, it's hardcoded as enp0s8

------

**FAQ**

**Q**: Will there be side-effect for "INTERFACES="ens32 ens192" ?
**A**: Nope. it will try to enable DHCP on both ens32 and ens192. in most case, they will NOT both exist.
but it will support both two NIC naming convention for e1000 & vmnet3. moreover, isc-dhcp-server will not fail even ens192 doesn't exist.

**Q**: why not change /etc/hosts together with hostname inside scripts/dep.sh ?
**A**: the /etc/hosts change requires reboot.  misalignment of run-time hostname and /etc/hosts will make lots of ```sudo``` operation very **slow** ( refer :http://askubuntu.com/questions/322514/terminal-command-with-sudo-takes-a-long-time). like firstuser.sh , the services start/stop will be very very slow. so move the /etc/hostname update into cleanup.sh

------

**Test Done**

After this PR fix, the new 16.04 OVA:
/etc/hostname will be rackhd
image
/etc/hosts will contain 127.0.1.1 rackhd
image

the isc-dhcp-server service runs normal.
image

-----

Thank you 
@yyscamper  @anhou  @cgx027  @changev 

---- vagrant box 16.04 Test result:

/etc/default/isc-dhcp-server
image

secondary NIC IP set to 172.31.128.1
image

service isc-dhcp-server status
image

### CODE REVIEW

**Comment by leoyjchang** [comment]
@panpan0000 Woo, very comprehensive content (what/how/test result/) provided which make the review easy and learnable. One small question, will isc-dhcp-server in Ubuntu 14.04 get impact with this new change?

File: packer/ansible/roles/isc-dhcp-server/files/systemd-isc-dhcp-server [hunk 1]
```diff
@@ -18,4 +18,5 @@
 
 # On what interfaces should the DHCP server (dhcpd) serve DHCP requests?
 #	Separate multiple interfaces with spaces, e.g. "eth0 eth1".
-INTERFACES="enp0s8"
+#   for 16.04, either ens33(e1000) or ens192(vmxnet3)
+INTERFACES="ens33 ens192"
```
Comment (cgx027) [inline packer/ansible/roles/isc-dhcp-server/files/systemd-isc-dhcp-server]: I am wondering if there are better ways to get the interface name than use the hard-cod-ed name. Any possibility that we inspect the syslog/dmesg and get the name?

```
onrack@~/github/on-static> dmesg | grep e1000
[    1.019679] e1000: Intel(R) PRO/1000 Network Driver - version 7.3.21-k8-NAPI
[    1.019680] e1000: Copyright (c) 1999-2006 Intel Corporation.
[    1.367837] e1000 0000:02:00.0 eth0: (PCI:66MHz:32-bit) 00:50:56:aa:bc:ab
[    1.367843] e1000 0000:02:00.0 eth0: Intel(R) PRO/1000 Network Connection
[    1.711464] e1000 0000:02:01.0 eth1: (PCI:66MHz:32-bit) 00:50:56:aa:66:16
[    1.711468] e1000 0000:02:01.0 eth1: Intel(R) PRO/1000 Network Connection
[    2.054803] e1000 0000:02:02.0 eth2: (PCI:66MHz:32-bit) 00:50:56:aa:11:76
[    2.054807] e1000 0000:02:02.0 eth2: Intel(R) PRO/1000 Network Connection
[    2.398718] e1000 0000:02:03.0 eth3: (PCI:66MHz:32-bit) 00:50:56:aa:8b:33
[    2.398723] e1000 0000:02:03.0 eth3: Intel(R) PRO/1000 Network Connection
[    5.054553] e1000: eth2 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
[    5.142130] e1000: eth3 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
[    5.163323] e1000: eth0 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
[    5.223507] e1000: eth1 NIC Link is Up 1000 Mbps Full Duplex, Flow Control: None
```

**Comment by panpan0000** [comment]
*****************************
Hold ON
*****************************
I soon realized this change will also impact vagrant box, I need to test against vagrant box.

I'm not sure about the 16.04 NIC name .

**Comment by panpan0000** [comment]
I found a new issue on Vagrant 16.04 building.
secondary NIC will not be populated in HW Config, unless speicified ```v.customize ["modifyvm", :id, "--nic2", "intnet"] ``` in Vagrantfile

I've tried add in template-ubuntu-16.04.json, but didn't work :-(
```
+      "vboxmanage_post": [
+        [ "modifyvm", "{{.Name}}", "--memory", "512" ],
+        [ "modifyvm", "{{.Name}}", "--cpus", "1" ],
+        [ "modifyvm", "{{.Name}}", "--nic2", "intnet" ]
       ]

```

File: packer/ansible/roles/isc-dhcp-server/files/systemd-isc-dhcp-server (hunk 1 above)
Comment (panpan0000) [inline packer/ansible/roles/isc-dhcp-server/files/systemd-isc-dhcp-server]: @cgx027 , I've finally use a "smart" secondary NIC name function to do what you proposed.

File: packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    #2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 ...
+    #.......
+    #3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pf....
+    #......
+    #
+    ########################
+    tmp=$( ip addr|grep "^${1}:" | awk '{print $2}')
+    if [[ "$tmp" == "" ]];
+    then
+        echo "[Error]: Invalid Parameter passed to get_nic_name_by_index(): NIC ID=$1";
```
Comment (cgx027) [inline packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh]: This sounds like verifying index parameter passed to ```get_nic_name_by_index```, but in a little bit tricky way. 

I suggest the best place to verify user input is before L23, if index is supposed to be a number, verify it against a number. 

Error catched here by should either be:
1. That's no interface number matching index parameter, maybe the index asked is to large.
2. Unfortunately (happens very rare), ```ip addr``` has changed it's output format thus the 2'nd awk field is no longer a valid interface name.

**Comment by cgx027** [comment]
@panpan0000 , amazing fast speed on the enhancement.

File: packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh (hunk 2 above)
Comment (panpan0000) [inline packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh]: 1. add numeric checking
2. add a check of the awk print output, to limit it in "xxx:" formating.

File: packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        exit 1
+    fi
+    name=${tmp/:/}  # remove the ":"
+    echo $name
+}
+
+#################################
+# retrieve secondary NIC Name(control port of RackHD)
+################################
+get_secondary_nic_name()
```
Comment (panpan0000) [inline packer/ansible/roles/isc-dhcp-server/files/config_isc-dhcp-server.sh]: it's a common function in my mind. not only used for control port of RackHD

File: packer/scripts/addnetif.sh [hunk 4]
```diff
@@ -10,28 +12,51 @@ setIP(){
     echo -e "\nauto $SECONDARY_NIC\niface $SECONDARY_NIC inet static\n   address $IP\n   netmask $MASK" >> /etc/network/interfaces
 }
 
+##########################################################
+# retrieve NIC name via ```ip addr``` command, by given id
+###########################################################
+get_nic_name_by_index()
```
Comment (panpan0000) [inline packer/scripts/addnetif.sh]: already put code in scripts/common folder.

**Comment by panpan0000** [comment]
@cgx027  @yyscamper  @anhou   , can you review it and approve ASAP ?
it's far away from the original purpose .... we need to fix 16.04 OVA isc-dhcp-server issue ASAP.

**Comment by anhou** [comment]
@panpan0000 Is it still WIP? seen from PR message title

**Comment by panpan0000** [comment]
@anhou , thanks, title fixed.

**Comment by anhou** [comment]
:+1:

**Pull request merged**